File: wpimath/src/test/native/cpp/geometry/struct/Pose2dStructTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <catch2/catch_approx.hpp>
#include <catch2/catch_test_macros.hpp>

#include "wpi/math/geometry/Pose2d.hpp"

using namespace wpi::math;

namespace {

using StructType = wpi::util::Struct<wpi::math::Pose2d>;
const Pose2d kExpectedData{
    Pose2d{Translation2d{0.191_m, 2.2_m}, Rotation2d{22.9_rad}}};
}  // namespace

TEST_CASE("Pose2dStructTest Roundtrip", "[wpimath]") {
  uint8_t buffer[StructType::GetSize()];
  std::memset(buffer, 0, StructType::GetSize());
  StructType::Pack(buffer, kExpectedData);

  Pose2d unpacked_data = StructType::Unpack(buffer);

  CHECK(kExpectedData.Translation() == unpacked_data.Translation());
  CHECK(kExpectedData.Rotation() == unpacked_data.Rotation());
}

File: wpimath/src/test/java/org/wpilib/math/geometry/struct/Pose2dStructTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.math.geometry.struct;

import static org.junit.jupiter.api.Assertions.assertEquals;

import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Translation2d;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import org.junit.jupiter.api.Test;

class Pose2dStructTest {
  private static final Pose2d DATA =
      new Pose2d(new Translation2d(0.191, 2.2), new Rotation2d(22.9));

  @Test
  void testRoundtrip() {
    ByteBuffer buffer = ByteBuffer.allocate(Pose2d.struct.getSize());
    buffer.order(ByteOrder.LITTLE_ENDIAN);
    Pose2d.struct.pack(buffer, DATA);
    buffer.rewind();

    Pose2d data = Pose2d.struct.unpack(buffer);
    assertEquals(DATA.getTranslation(), data.getTranslation());
    assertEquals(DATA.getRotation(), data.getRotation());
  }
}
